File: 20-actors/kanjo/methods/autorun.py
#!/usr/bin/env python3
"""autorun.py — kanjō AUTONOMOUS financial-disclosure heartbeat on the kotoba Datom log.
ADR-2606032000.

This is what "kotoba で自律的に稼働する" means for kanjō, in the constitution-permitted form
(mirrors shionome / ipaddress / yabai / sukashi / watatsuna / watari / kabuto autorun). Each
heartbeat the actor runs its whole disclosure pipeline ITSELF, with no human in the loop:

    observe (load the OFFLINE merged disclosed-fact graph) → split filings / facts
      → by-company-year → derive ratios + YoY (:synthesized) → sector/currency aggregates
        (coverage-honest, no cross-currency FX sums)
      → PERSIST a content-addressed transaction to the append-only kotoba Datom log
        (graph datoms + derived :fin.metric + :fin.agg), linking the previous tx's CID.

Constitutional posture holds by construction: only disclosed primary-filing FACTS + transparent
ratios are representable — never a rating, valuation, solvency verdict, FORECAST, or buy/sell call
(G2/G4); derived metrics/aggregates carry :sourcing :synthesized and are NEVER re-ingested as
disclosed facts (G5); a restatement is a new fact + :superseded-by, never a deletion (G11, 非終末論).

The loop is deterministic / resume-safe (cycle drives tx-id + as-of → same CIDs; the derived path
uses no PYTHONHASHSEED-randomized set iteration — by-company-year + metric-inputs are dict-ordered,
aggregates are sorted) and append-only. WHAT STAYS GATED (G7): it NEVER fetches live EDGAR/EDINET
and NEVER pushes to a live kotoba node. Ingest is the offline merged/seed graph; persistence is the
LOCAL append-only log. Live universe ingest is Council + operator gated. Stdlib only.
"""
from __future__ import annotations

import argparse
import os
import pathlib
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import kanjo_edn  # noqa: E402
from analyze import aggregates, by_company_year, derive_metrics  # noqa: E402
from kotoba import (append_tx, derived_datoms, graph_datoms, head_cid,  # noqa: E402
                    make_tx, read_log, verify_chain)

HERE = pathlib.Path(__file__).resolve().parent
DATA = HERE.parent / "data"
MERGED = DATA / "facts.merged.kotoba.edn"
SEED = DATA / "seed-financial-facts.kotoba.edn"
LOG = DATA / "kanjo.datoms.kotoba.edn"
BASE_AS_OF = 20260609


def _graph_path(graph_path: pathlib.Path | None) -> pathlib.Path:
    if graph_path is not None:
        return graph_path
    return MERGED if MERGED.exists() else SEED


def run_cycle(cycle: int, graph_path: pathlib.Path | None = None,
              log_path: pathlib.Path = LOG) -> dict:
    """One autonomous heartbeat: observe → derive ratios/YoY + aggregates → persist a
    content-addressed Datom transaction (graph + derived :fin.metric + :fin.agg). cycle drives
    tx-id + as-of."""
    rows = kanjo_edn.read_file(str(_graph_path(graph_path)))   # observe — OFFLINE (G7: no live fetch)
    facts = [r for r in rows if isinstance(r, dict) and ":fin.fact/id" in r]
    filings = [r for r in rows if isinstance(r, dict) and ":fin.filing/id" in r]
    cy = by_company_year(facts)
    metrics = derive_metrics(cy)                  # :synthesized ratios + YoY (G5)
    aggs = aggregates(cy)                          # coverage-honest sector/currency aggregates
    datoms = graph_datoms(rows) + derived_datoms(metrics, aggs)
    tx = make_tx(datoms, tx_id=cycle, as_of=BASE_AS_OF + cycle, prev_cid=head_cid(log_path))
    cid = append_tx(tx, log_path)                  # PERSIST to append-only LOCAL kotoba log
    return {
        "cycle": cycle,
        "filings": len(filings),
        "facts": len(facts),
        "companies": len(cy),
        "metrics": len(metrics),
        "aggregates": len(aggs),
        "datoms": len(datoms),
        "cid": cid,
    }


def run_autonomous(cycles: int = 3, graph_path: pathlib.Path | None = None,
                   log_path: pathlib.Path = LOG) -> dict:
    beats = [run_cycle(c, graph_path, log_path) for c in range(1, cycles + 1)]
    return {
        "cycles": cycles,
        "beats": beats,
        "log_length": len(read_log(log_path)),
        "head_cid": head_cid(log_path),
        "chain": verify_chain(log_path),
    }


if __name__ == "__main__":
    ap = argparse.ArgumentParser(description="kanjo autonomous financial-disclosure loop")
    ap.add_argument("--cycles", type=int, default=3, help="number of self-paced heartbeats")
    ap.add_argument("--graph", type=pathlib.Path, default=None, help="disclosed-fact graph EDN (offline)")
    ap.add_argument("--log", type=pathlib.Path, default=LOG, help="kotoba Datom log path")
    ap.add_argument("--fresh", action="store_true", help="start a fresh log (remove existing)")
    args = ap.parse_args()
    if args.fresh and args.log.exists():
        args.log.unlink()
    res = run_autonomous(args.cycles, graph_path=args.graph, log_path=args.log)
    print("# kanjo — AUTONOMOUS financial-disclosure over the kotoba Datom log "
          "(offline ingest, LOCAL persist; live EDGAR/EDINET / live-node push stays G7-gated)\n")
    for bt in res["beats"]:
        print(f"  ♥ cycle {bt['cycle']}: {bt['filings']} filings / {bt['facts']} facts / "
              f"{bt['companies']} companies · metrics {bt['metrics']} · aggregates {bt['aggregates']} "
              f"+{bt['datoms']} datoms → cid {bt['cid'][:14]}…")
    ch = res["chain"]
    print(f"\n  log: {res['log_length']} tx · head {res['head_cid'][:14]}… · "
          f"chain {'OK ✓' if ch['ok'] else 'BROKEN at ' + str(ch['broken_at'])} · "
          f"disclosed facts + :synthesized ratios, non-adjudicating / no forecast (G2/G4)")

File: 20-actors/kanjo/methods/autorun.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/autorun.py — the autonomous financial-disclosure heartbeat.
(ns kanjo.methods.autorun
  "kanjō 勘定 — AUTONOMOUS financial-disclosure heartbeat on the kotoba Datom log (ADR-2606032000).

  Each heartbeat: observe (OFFLINE merged/seed graph, G7 no live feed) → split filings/facts →
  by-company-year → derive ratios + YoY + sector/currency aggregates → PERSIST a content-addressed
  transaction (graph datoms + derived :fin.metric + :fin.agg) linking the prev CID. Deterministic /
  resume-safe; no external I/O; live EDGAR/EDINET push stays G7-gated. Non-adjudicating / no-advice /
  no-forecast (G2/G4); every derived datom :sourcing :synthesized (G5).

  Run:  bb --classpath 20-actors 20-actors/kanjo/methods/autorun.clj --cycles 3 --fresh"
  (:require [kanjo.methods.kanjo-edn :as ke]
            [kanjo.methods.analyze :as a]
            [kanjo.methods.kotoba :as k]
            [clojure.java.io :as io]
            [clojure.string :as str]))

(def ^:private this-file *file*)
(defn- data-dir [] (-> this-file io/file .getAbsoluteFile .getParentFile .getParentFile (io/file "data")))
(defn- merged [] (io/file (data-dir) "facts.merged.kotoba.edn"))
(defn- seed [] (io/file (data-dir) "seed-financial-facts.kotoba.edn"))
(defn- default-log [] (io/file (data-dir) "kanjo.datoms.kotoba.edn"))
(def BASE-AS-OF 20260608)

(defn- graph-path [gp] (or gp (let [m (merged)] (if (.exists m) m (seed)))))

(defn run-cycle
  [cycle & {:keys [graph-path* log-path]}]
  (let [log-path (or log-path (default-log))
        gp (graph-path graph-path*)
        rows (ke/read-file gp)
        [filings facts] (a/load gp)
        cy (a/by-company-year facts)
        metrics (a/derive-metrics cy)
        aggs (a/aggregates cy)
        datoms (vec (concat (k/graph-datoms rows) (k/derived-datoms metrics aggs)))
        tx (k/make-tx datoms :tx-id cycle :as-of (+ BASE-AS-OF cycle) :prev-cid (k/head-cid log-path))
        cid (k/append-tx tx log-path)]
    {:cycle cycle :filings (count filings) :facts (count facts) :companies (count cy)
     :metrics (count metrics) :aggregates (count aggs) :datoms (count datoms) :cid cid}))

(defn run-autonomous [& {:keys [cycles graph-path* log-path] :or {cycles 3}}]
  (let [log-path (or log-path (default-log))
        beats (mapv #(run-cycle % :graph-path* graph-path* :log-path log-path) (range 1 (inc cycles)))]
    {:cycles cycles :beats beats :log-length (count (k/read-log log-path))
     :head-cid (k/head-cid log-path) :chain (k/verify-chain log-path)}))

(defn -main [& argv]
  (let [args (vec argv)
        cyc-idx (.indexOf args "--cycles")
        cycles (if (>= cyc-idx 0) (Integer/parseInt (nth args (inc cyc-idx))) 3)
        log-idx (.indexOf args "--log")
        log-path (if (>= log-idx 0) (io/file (nth args (inc log-idx))) (default-log))]
    (when (and (some #{"--fresh"} args) (.exists (io/file log-path))) (.delete (io/file log-path)))
    (let [res (run-autonomous :cycles cycles :log-path log-path)]
      (println (str "# kanjō — AUTONOMOUS financial-disclosure observation over the kotoba Datom log "
                    "(offline ingest, LOCAL persist; live EDGAR/EDINET stays G7-gated)\n"))
      (doseq [bt (:beats res)]
        (println (format "  ♥ cycle %d: %d filings / %d facts / %d cos · %d metrics · %d aggregates +%d datoms → cid %s…"
                         (:cycle bt) (:filings bt) (:facts bt) (:companies bt) (:metrics bt)
                         (:aggregates bt) (:datoms bt) (subs (:cid bt) 0 14))))
      (let [ch (:chain res)]
        (println (format "\n  log: %d tx · head %s… · chain %s · disclosed facts + :synthesized ratios, never a verdict (G2/G4/G5)"
                         (:log-length res) (subs (:head-cid res) 0 14)
                         (if (:ok ch) "OK ✓" (str "BROKEN at " (:broken-at ch)))))))))

(when (= *file* (System/getProperty "babashka.file"))
  (apply -main *command-line-args*))
